python/data_structures/binary_tree.py

```diff
@@ -48,3 +48,15 @@ def check_nodes(root, values):
             return values
         check_nodes(self.root, values)
         return values
+
+    def find_maximum_value(self):
+        maxv = self.root.value
+        def check_nodes(root):
+            nonlocal maxv
+            if not root:
+                return
+            maxv = max(root.value, maxv)
+            check_nodes(root.left)
+            check_nodes(root.right)
+            return maxv
+        return check_nodes(self.root)
```

Comment:
You've got kind of a mix of two different approaches here. If you wanted to trim up the logic a bit, there are two possible refactors I would suggest. Either:

1. Remove line 61, `return maxv` and just return `maxv` on line 62. Since you're utilizing `nonlocal` the return is kind of unnecessary.

or

2. Remove line 55, accept `maxv` as an arg of `check_nodes(root, maxv)`, and unsure you capture the returns on 59 and 60 `maxv = check_nodes(root.left, maxv)`

Not super necessary, but I figured I would point it out. Option 2 is more of a "pure function" approach since it avoids any side-effects outside of the check_nodes helper function.